sdk/scheduler/src/main/java/com/mesosphere/sdk/debug/PlansTracker.java

```diff
@@ -0,0 +1,297 @@
+      outcome.put("invalid_input",
+          "Supplied plan not found in list of all available plans!");
+      return Optional.of(outcome);
+    }
+
+    //If no explicit phase defined, nothing further to do.
+    if (filterPhase == null) {
+      return Optional.empty();
+    }
+
+    Plan plan = planManagers.get(0).getPlan();
+
+    //Ensure our phase is found in the plan.
+    List<Phase> phaseList = plan.getChildren().stream()
+        .filter(phases -> phases.getName().equalsIgnoreCase(filterPhase))
+        .collect(Collectors.toList());
+
+    //Ensure we found our phase.
+    if (phaseList.size() != 1) {
+      JSONObject outcome = new JSONObject();
+      outcome.put("invalid_input",
+          "Supplied phase not found in set of possible phases with supplied plan!");
+      return Optional.of(outcome);
+    }
+
+    //If no explicit step defined, nothing further to do.
+    if (filterStep == null) {
+      return Optional.empty();
+    }
+
+    Phase phase = phaseList.get(0);
+    List<Step> stepList = phase.getChildren().stream()
+        .filter(steps -> steps.getName().equalsIgnoreCase(filterStep))
+        .collect(Collectors.toList());
+
+    //Ensure we found our phase.
+    if (stepList.size() != 1) {
+      JSONObject outcome = new JSONObject();
+      outcome.put("invalid_input",
+          "Supplied step not found in set of possible steps with supplied plan and phase!");
+      return Optional.of(outcome);
+    }
+
+    //Successfully found plan, phase and step.
+    return Optional.empty();
+  }
+
+  public Response getJson(@QueryParam("plan") String filterPlan,
+                        @QueryParam("phase") String filterPhase,
+                        @QueryParam("step") String filterStep,
+                        @QueryParam("sync") boolean requireSync)
+  {
+
+    //Validate plan/phase/step if provided.
+    if (filterPlan != null || filterPhase != null || filterStep != null) {
+      Optional<JSONObject> validationOutcome = getValidationErrorResponse(filterPlan,
+                                                                          filterPhase,
+                                                                          filterStep);
+      if (validationOutcome.isPresent()) {
+        return ResponseUtils.jsonOkResponse(validationOutcome.get());
+      }
+    }
+
+    //At this point we're either returning the entire plans tree or
+    //pruning it down to a plan/phase/step which has been validated.
+
+    JSONObject response = new JSONObject();
+    response.put("service-topology", generateServiceStatus(filterPlan,
+        filterPhase,
+        filterStep,
+        false,
+        true));
+
+    response.put("service-status", generateServiceStatus(filterPlan,
+        filterPhase,
+        filterStep,
+        true,
+        false));
+
+    //Retrieve the latest Plans from PlanCoordinator. Some plans i.e recovery do change over time.
+    HashMap<String, Plan> planMap = new HashMap<String, Plan>();
```

Comment:
`HashMap<String, Plan> planMap = new HashMap<String, Plan>(` could be `HashMap<String, Plan> planMap = new HashMap<>(`